pkg/aspect/metricsManager.go

```diff
@@ -0,0 +1,227 @@
+	var values []adapter.Value
+
+	for name, md := range w.metadata {
+		metricValue, err := mapper.Eval(md.value, attrs)
+		if err != nil {
+			result = multierror.Append(result, fmt.Errorf("failed to eval metric value for metric '%s' with err: %s", name, err))
+			continue
+		}
+		labels, err := evalAll(md.labels, attrs, mapper)
+		if err != nil {
+			result = multierror.Append(result, fmt.Errorf("failed to eval labels for metric '%s' with err: %s", name, err))
+			continue
+		}
+
+		// TODO: investigate either pooling these, or keeping a set around that has only its field's values updated.
+		// we could keep a map[metric name]value, iterate over the it updating only the fields in each value
+		values = append(values, adapter.Value{
+			Name:   name,
+			Kind:   md.metricKind,
+			Labels: labels,
+			// TODO: extract standard timestamp attributes for start/end once we det'm what they are
+			StartTime:   time.Now(),
+			EndTime:     time.Now(),
+			MetricValue: metricValue,
+		})
+	}
+
+	if err := w.aspect.Record(values); err != nil {
+		result = multierror.Append(result, fmt.Errorf("failed to record all values with err: %s", err))
+	}
+
+	// We may accumulate errors while still being able to record some metrics; if we have any values at all we'll
+	// return an OK alongside our errors (since presumably we're able to record at least a few).
+	var out *Output
+	if len(values) > 0 {
+		out = &Output{Code: code.Code_OK}
+	}
+	return out, result.ErrorOrNil()
+}
+
+func (w *metricsWrapper) Close() error {
+	return w.aspect.Close()
+}
+
+func evalAll(expressions map[string]string, attrs attribute.Bag, eval expr.Evaluator) (map[string]interface{}, error) {
+	result := &multierror.Error{}
+	labels := make(map[string]interface{}, len(expressions))
+	for label, texpr := range expressions {
+		val, err := eval.Eval(texpr, attrs)
+		if err != nil {
+			result = multierror.Append(result, fmt.Errorf("failed to construct value for label '%s' with err: %s", label, err))
+			continue
+		}
+		labels[label] = val
+	}
+	return labels, result.ErrorOrNil()
+}
+
+func definitionFromProto(desc *dpb.MetricDescriptor) (*adapter.MetricDefinition, error) {
+	labels := make(map[string]adapter.LabelType, len(desc.Labels))
+	for _, label := range desc.Labels {
+		l, err := adapter.LabelTypeFromProto(label.ValueType)
+		if err != nil {
+			return nil, fmt.Errorf("descriptor '%s' label '%s' failed to convert label type value '%v' from proto with err: %s",
+				desc.Name, label.Name, label.ValueType, err)
+		}
+		labels[label.Name] = l
+	}
+	kind, err := adapter.MetricKindFromProto(desc.Kind)
+	if err != nil {
+		return nil, fmt.Errorf("descriptor '%s' failed to convert metric kind value '%v' from proto with err: %s",
+			desc.Name, desc.Kind, err)
+	}
+	return &adapter.MetricDefinition{
+		Name:   desc.Name,
+		Kind:   kind,
+		Labels: labels,
+	}, nil
+}
+
+func find(defs []*aconfig.MetricsParams_Metric, name string) (*aconfig.MetricsParams_Metric, bool) {
```

Comment:
aside: at some point, i imagine we will want a version of this that takes a fn and can be used more broadly across other slices, etc.